File: avl.py
import sys
import random

# Create tree node
class TreeNode(object):
    def __init__(node, key):
        node.key = key
        node.left = None
        node.right = None
        node.height = 1


class AVLTree(object):

    # Function to insert a node
    def insert_node(node, root, key):

        # Find the correct location and insert the node
        if not root:
            return TreeNode(key)
        elif key < root.key:
            root.left = node.insert_node(root.left, key)
        else:
            root.right = node.insert_node(root.right, key)

        root.height = 1 + max(node.getHeight(root.left),
                              node.getHeight(root.right))

        # Update the balance factor and balance the tree
        balanceFactor = node.getBalance(root)
        if balanceFactor > 1:
            if key < root.left.key:
                return node.rightRotate(root)
            else:
                root.left = node.leftRotate(root.left)
                return node.rightRotate(root)

        if balanceFactor < -1:
            if key > root.right.key:
                return node.leftRotate(root)
            else:
                root.right = node.rightRotate(root.right)
                return node.leftRotate(root)

        return root

    # Function to delete a node
    def delete_node(node, root, key):

        # Find the node to be deleted and remove it
        if not root:
            return root
        elif key < root.key:
            root.left = node.delete_node(root.left, key)
        elif key > root.key:
            root.right = node.delete_node(root.right, key)
        else:
            if root.left is None:
                temp = root.right
                root = None
                return temp
            elif root.right is None:
                temp = root.left
                root = None
                return temp
            temp = node.getMinValueNode(root.right)
            root.key = temp.key
            root.right = node.delete_node(root.right,
                                          temp.key)
        if root is None:
            return root

        # Update the balance factor of nodes
        root.height = 1 + max(node.getHeight(root.left),
                              node.getHeight(root.right))

        balanceFactor = node.getBalance(root)

        # Balance the tree
        if balanceFactor > 1:
            if node.getBalance(root.left) >= 0:
                return node.rightRotate(root)
            else:
                root.left = node.leftRotate(root.left)
                return node.rightRotate(root)
        if balanceFactor < -1:
            if node.getBalance(root.right) <= 0:
                return node.leftRotate(root)
            else:
                root.right = node.rightRotate(root.right)
                return node.leftRotate(root)
        return root

    # Function to perform left rotation
    def leftRotate(node, z):
        y = z.right
        T2 = y.left
        y.left = z
        z.right = T2
        z.height = 1 + max(node.getHeight(z.left),
                           node.getHeight(z.right))
        y.height = 1 + max(node.getHeight(y.left),
                           node.getHeight(y.right))
        return y

    # Function to perform right rotation
    def rightRotate(node, z):
        y = z.left
        T3 = y.right
        y.right = z
        z.left = T3
        z.height = 1 + max(node.getHeight(z.left),
                           node.getHeight(z.right))
        y.height = 1 + max(node.getHeight(y.left),
                           node.getHeight(y.right))
        return y

    # Get the height of the node
    def getHeight(node, root):
        if not root:
            return 0
        return root.height

    # Get balance factore of the node
    def getBalance(node, root):
        if not root:
            return 0
        return node.getHeight(root.left) - node.getHeight(root.right)

    def getMinValueNode(node, root):
        if root is None or root.left is None:
            return root
        return node.getMinValueNode(root.left)

    def preOrder(node, root):
        if not root:
            return
        print("{0} ".format(root.key), end="")
        node.preOrder(root.left)
        node.preOrder(root.right)

    # Print the tree
    def printHelper(node, currPtr, indent, last):
        if currPtr != None:
            sys.stdout.write(indent)
            if last:
                sys.stdout.write("R----")
                indent += "     "
            else:
                sys.stdout.write("L----")
                indent += "|    "
            print(currPtr.key)
            node.printHelper(currPtr.left, indent, False)
            node.printHelper(currPtr.right, indent, True)


myTree = AVLTree()
root = None
nums = random.sample(range(0, 100), 8)
for num in nums:
    root = myTree.insert_node(root, num)
myTree.printHelper(root, "", True)
key = nums[7]
root = myTree.delete_node(root, key)
print("After Deletion: ")
myTree.printHelper(root, "", True)

File: avl.java
import java.util.*;

class Node 
{
	int item, height;
	Node left, right;

	Node(int d) 
	{
		item = d;
		height = 1;
	}
}

class AVLTree 
{
	Node root;

	// returns tree height
	int height(Node N) 
	{
		if (N == null)
			return 0;
		return N.height;
	}

	// returns the larger of the inputed values
	int max(int a, int b) 
	{
		if (a > b)
		{
			return a;
		}
		return b;
	}

	// rotates subtree right
	Node rightRotate(Node y) 
	{
		Node x = y.left;
		Node T2 = x.right;
		x.right = y;
		y.left = T2;
		y.height = max(height(y.left), height(y.right)) + 1;
		x.height = max(height(x.left), height(x.right)) + 1;
		return x;
	}

	// rotates subtree left
	Node leftRotate(Node x) 
	{
		Node y = x.right;
		Node T2 = y.left;
		y.left = x;
		x.right = T2;
		x.height = max(height(x.left), height(x.right)) + 1;
		y.height = max(height(y.left), height(y.right)) + 1;
		return y;
	}

	// get balance factor of a node
	int getBalanceFactor(Node N) 
	{
		if (N == null)
		{
			return 0;
		}
		return height(N.left) - height(N.right);
	}

	// insert a node
	Node insertNode(Node node, int item) 
	{
		// insert
		if (node == null)
		{
			return (new Node(item));
		}
		if (item < node.item)
		{
			node.left = insertNode(node.left, item);
		}
		else if (item > node.item)
		{
			node.right = insertNode(node.right, item);
		}
		else
		{
			return node;
		}

		// update the balance factor of nodes and balance tree
		node.height = 1 + max(height(node.left), height(node.right));
		int balanceFactor = getBalanceFactor(node);
		if (balanceFactor > 1) 
		{
			if (item < node.left.item) 
			{
				return rightRotate(node);
			} 
			else if (item > node.left.item) 
			{
				node.left = leftRotate(node.left);
				return rightRotate(node);
			}
		}
		if (balanceFactor < -1) 
		{
			if (item > node.right.item) 
			{
				return leftRotate(node);
			} 
			else if (item < node.right.item) 
			{
				node.right = rightRotate(node.right);
				return leftRotate(node);
			}
		}
		return node;
	}

	// returns node with smallest value   
	Node nodeWithSmallestValue(Node node) 
	{
		Node curr = node;
		while (curr.left != null)
		{
			curr = curr.left;
		}
		return curr;
	}

	// Delete a node
	Node deleteNode(Node root, int item) 
	{
		// Find the node to be deleted and remove it
		if (root == null)
		{
			return root;
		}
		if (item < root.item)
		{
			root.left = deleteNode(root.left, item);
		}
		else if (item > root.item)
		{
			root.right = deleteNode(root.right, item);
		}
		else 
		{
			if ((root.left == null) || (root.right == null)) 
			{
				Node temp = null;
				if (temp == root.left)
				{
					temp = root.right;
				}	
				else
				{
					temp = root.left;
				}
				if (temp == null) 
				{
					temp = root;
					root = null;
				} 
				else
				{
					root = temp;
				}	
			} 
			else 
			{
				Node temp = nodeWithSmallestValue(root.right);
				root.item = temp.item;
				root.right = deleteNode(root.right, temp.item);
			}
		}
		if (root == null)
		{
			return root;
		}

		// update the balance factor of each node and balance the tree
		root.height = max(height(root.left), height(root.right)) + 1;
		int balanceFactor = getBalanceFactor(root);
		if (balanceFactor > 1) 
		{
			if (getBalanceFactor(root.left) >= 0) 
			{
				return rightRotate(root);
			} 
			else 
			{
				root.left = leftRotate(root.left);
				return rightRotate(root);
			}
		}
		if (balanceFactor < -1) 
		{
			if (getBalanceFactor(root.right) <= 0) 
			{
				return leftRotate(root);
			} 
			else 
			{
				root.right = rightRotate(root.right);
				return leftRotate(root);
			}
		}
		return root;
	}

	void preOrder(Node node) 
	{
		if (node != null) 
		{
			System.out.print(node.item + " ");
			preOrder(node.left);
			preOrder(node.right);
		}
	}

	// print tree
	private void printTree(Node curr, String space, boolean last) 
	{
		if (curr != null) 
		{
			System.out.print(space);
			if (last) 
			{
				System.out.print("R----");
				space += "   ";
			} 
			else 
			{
				System.out.print("L----");
				space += "|  ";
			}
			System.out.println(curr.item);
			printTree(curr.left, space, false);
			printTree(curr.right, space, true);
		}
	}

	// driver code
	public static void main(String[] args) 
	{
		Random rand = new Random();
		AVLTree tree = new AVLTree();
		int r;
		for (int i = 0; i < 7; i++)
		{
			r = rand.nextInt(100);
			tree.root = tree.insertNode(r);
		}
		tree.printTree(tree.root, "", true);
		tree.root = tree.deleteNode(tree.root, r);
		System.out.println("After Deletion of " + r +": ");
		tree.printTree(tree.root, "", true);
	}
}